flink-libraries/flink-table/src/main/scala/org/apache/flink/table/functions/utils/UserDefinedFunctionUtils.scala

```diff
@@ -314,7 +314,28 @@ object UserDefinedFunctionUtils {
       aggregateFunction: AggregateFunction[_, _],
       extractedType: TypeInformation[_] = null)
     : TypeInformation[_] = {
-    getParameterTypeOfAggregateFunction(aggregateFunction, "getResultType", 0, extractedType)
+
+    val resultType = aggregateFunction.getResultType
+    if (resultType != null) {
+      resultType
+    } else if (extractedType != null) {
+      extractedType
+    } else {
+      try {
+        TypeExtractor
+          .createTypeInfo(aggregateFunction,
+            classOf[AggregateFunction[_, _]],
+            aggregateFunction.getClass,
+            0)
+          .asInstanceOf[TypeInformation[_]]
+      } catch {
+        case ite: InvalidTypesException =>
+          throw new TableException(
+            s"Cannot infer generic type of ${aggregateFunction.getClass}. " +
+              s"You can override AggregateFunction.getResultType() to specify the type.",
+            ite)
+      }
+    }
```

Comment:
Just  a suggestion(I am not sure the suggestion is better or not):
Extract common code for `getResultTypeOfAggregateFunction` and `getAccumulatorTypeOfAggregateFunction`. Something like following:
![image](https://user-images.githubusercontent.com/22488084/28444428-c85f8dc2-6def-11e7-996e-de80a728a6ce.png)